aten/src/ATen/templates/Operators.h

```diff
@@ -24,23 +27,77 @@
-#define ATEN_FN(op_name) at::_ops::op_name
+#define ATEN_FN2(op_name, overload) at::_ops::op_name##_##overload::call
+#define ATEN_FN(op_name) at::_ops::op_name::call
+
+// Separately, ATEN_OP(op) and ATEN_OP2(op, overload) define a class containing compile-time
+// metadata about a given aten operator.
+// Notable data on the class includes:
+// - ATEN_OP2(add, Tensor)::name // returns the string name: "add"
+// - ATEN_OP2(add, Tensor)::overload_name // returns the string overload name: "Tensor"
+// - ATEN_OP2(add, Tensor)::schema // returns the C++ schema type: at::Tensor (const at::Tensor &, const at::Tensor &, const at::Scalar &)
+// - ATEN_OP2(add, Tensor)::schema_str // returns the string jit type: "add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor"
+
+#define ATEN_OP2(op_name, overload) at::_ops::op_name##_##overload
+#define ATEN_OP(op_name) at::_ops::op_name
 
 // WARNING: Please do not call any of the ops in the _ops namespace directly.
 // Use the ATEN_FN macros. We do not guarantee stability of the naming
 // scheme for the functions in at::_ops
-namespace at { namespace _ops {
 
-// NB: We are forced to special case requires_grad_. This is because all
-// of the auto-generated inplace method signatures in TensorMethods.h are
-// codegen'ed to return Tensor&, but requires_grad_ has a `manual_cpp_binding`
-// with a different signature that returns `const Tensor&`.
-//
-// Eventually, the plan is to kill Tensor& from all C++ signatures and use
-// const Tensor&. When that happens, we can remove this special case and just
-// let the codegen handle it.
-TORCH_API Tensor & requires_grad_(Tensor & self, bool requires_grad);
+// See Note [The ATen Operators API] for details of the at::_ops namespace
+
+namespace c10 { namespace impl {
+
+inline c10::optional<MemoryFormat>
+check_tensor_options_and_extract_memory_format(
+    const TensorOptions& options,
+    c10::optional<MemoryFormat> memory_format) {
+  TORCH_CHECK(
+      options.requires_grad_opt() == c10::nullopt ||
+          options.requires_grad_opt().value() == false,
+      "Operators taking TensorOptions cannot take a TensorOptions with "
+      "options.requires_grad set as true. This isn't implemented yet.");
+  TORCH_CHECK(
+      !(options.has_memory_format() && memory_format.has_value()),
+      "Cannot set memory_format both in TensorOptions and explicit argument; please delete "
+      "the redundant setter.");
+  if (memory_format.has_value()) {
+    return memory_format;
+  } else {
+    return options.memory_format_opt();
+  }
+}
+
+}} // namespace impl namespace c10
+
+
+// Forward declarations of any types needed in the operator signatures.
+// We can't directly include these classes because it will cause circular include dependencies.
+// This file is included by TensorBody.h, which defines the Tensor class.
+namespace c10 {
+
+template<typename T>
+class optional;
+template<typename T>
+class List;
+class Stream;
+struct Storage;
```

Comment:
forward declared a bunch of classes needed in the function declarations; We don't need to #include any of these files since we're only declaring method signatures in this file.